### PR TITLE
add lint for inline asm labels that look like binary

### LINKED ISSUE #94426: in asm!() using a local numeric label made of all 0's and 1's gives a confusing error
```
rustc 1.59.0 (9d1b2106e 2022-02-23)
```

Here's a function that jumps to a local numberic label inside of `asm!`:

```rust
use std::arch::asm;

#[cfg(target_arch = "x86_64")]
fn main() {
    let mut counter: u64 = 0;
    unsafe {
        asm!(
            "2:",
            "inc {counter}",
            "cmp {counter}, 10",
            "jnz 2b",
            counter = inout(reg) counter,
        );
    }
    dbg!(counter);  // prints 10
}
```

As per [Rust By Example](https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels), we use `2:` as a local label there to work around an LLVM bug that mistakes labels like `0:`, `1:`, or `10:` for binary. This example works. However, if we don't know about this rule, and we try to use `0:` instead, the error we get is confusing:

```
error: invalid operand for instruction
  --> src/main.rs:11:14
   |
11 |             "jnz 0b",
   |              ^
   |
note: instantiated into assembly here
  --> <inline asm>:5:1
   |
5  | jnz 0b
   | ^

error: could not compile `scratch` due to previous error
```

Using `1:` is the same:

```
error: invalid operand for instruction
  --> src/main.rs:11:14
   |
11 |             "jnz 1b",
   |              ^
   |
note: instantiated into assembly here
  --> <inline asm>:5:1
   |
5  | jnz 1b
   | ^

error: could not compile `scratch` due to previous error
```

It would be pretty great to have a hint here like "`0:` and `1:` aren't valid local labels unfortunately, so just start with `2:`".

### PR DESCRIPTION
**Pull request by asquared31415**
fixes #94426

Due to a bug/feature in LLVM, labels composed of only the digits `0` and `1` can sometimes be confused with binary literals, even if a binary literal would not be valid in that position.

This PR adds detection for such labels and also as a drive-by change, adds a note to cases such as `asm!(include_str!("file"))` that the label that it found came from an expansion of a macro, it wasn't found in the source code.

I expect this PR to upset some people that were using labels `0:` or `1:` without issue because they never hit the case where LLVM got it wrong, but adding a heuristic to the lint to prevent this is not feasible - it would involve writing a whole assembly parser for every target that we have assembly support for.

[zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-06-20/near/445870628)

r? @estebank

### CODE REVIEW

**Comment by asquared31415** [comment]
cc @Amanieu

**Comment by joshtriplett** [comment]
This seems like a good starting point! 

Eventually, I hope we can get LLVM to have a mode without this "feature", and then rust assembly should use that mode by default. But in the interim, we should warn people.

File: compiler/rustc_lint/src/builtin.rs [hunk 1]
```diff
@@ -2804,10 +2803,48 @@ declare_lint! {
...
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// A bug in LLVM causes this code to fail
+    ///
+    /// See the explanation in [Rust By Example] for more details.
+    ///
+    /// [local labels]: https://sourceware.org/binutils/docs/as/Symbol-Names.html#Local-Labels
+    /// [Rust By Example]: https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels
```
Comment (estebank) [inline compiler/rustc_lint/src/builtin.rs:2832]: We don't currently explain this issue there, do we?

File: compiler/rustc_lint/src/builtin.rs [hunk 2]
```diff
@@ -2804,10 +2803,48 @@ declare_lint! {
...
+    ///         asm!("0: bar");
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// A bug in LLVM causes this code to fail
```
Comment (estebank) [inline compiler/rustc_lint/src/builtin.rs:2827]: I would like us to expand on this explanation a bit, and ideally link to the original bug tracker here, unless we start to do so from rust by example.

File: compiler/rustc_lint/src/builtin.rs [hunk 3]
```diff
@@ -2858,16 +2896,21 @@ impl<'tcx> LateLintPass<'tcx> for NamedAsmLabels {
...
-                        // A label starts with an ASCII alphabetic character or . or _
                         // A label can also start with a format arg, if it's not a raw asm block.
                         if !raw && start == '{' {
                             in_bracket = true;
+                            label_kind = AsmLabelKind::FormatArg;
+                        } else if matches!(start, '0' | '1') {
+                            // binary labels have only the characters 0 or 1
+                            label_kind = AsmLabelKind::Binary;
                         } else if !(start.is_ascii_alphabetic() || matches!(start, '.' | '_')) {
+                            // named labels start with ASCII letters or . or _
```
Comment (estebank) [inline compiler/rustc_lint/src/builtin.rs:2909]:
```suggestion
                            // Named labels start with ASCII letters, `.` or `_`.
```

File: compiler/rustc_lint/src/builtin.rs [hunk 4]
```diff
@@ -2858,16 +2896,21 @@ impl<'tcx> LateLintPass<'tcx> for NamedAsmLabels {
...
                         let mut in_bracket = false;
+                        let mut label_kind = AsmLabelKind::Named;
 
-                        // A label starts with an ASCII alphabetic character or . or _
                         // A label can also start with a format arg, if it's not a raw asm block.
                         if !raw && start == '{' {
                             in_bracket = true;
+                            label_kind = AsmLabelKind::FormatArg;
+                        } else if matches!(start, '0' | '1') {
+                            // binary labels have only the characters 0 or 1
```
Comment (estebank) [inline compiler/rustc_lint/src/builtin.rs:2906]:
```suggestion
                            // Binary labels have only the characters `0` or `1`.
```

File: compiler/rustc_lint/src/builtin.rs [hunk 5]
```diff
@@ -2888,34 +2931,55 @@ impl<'tcx> LateLintPass<'tcx> for NamedAsmLabels {
                             } else if !raw && c == '{' {
                                 // Start of a format arg.
                                 in_bracket = true;
+                                label_kind = AsmLabelKind::FormatArg;
                             } else {
-                                if !(c.is_ascii_alphanumeric() || matches!(c, '_' | '$')) {
+                                let can_continue = match label_kind {
+                                    // format arg labels are considered to be named labels for the purposes
```
Comment (estebank) [inline compiler/rustc_lint/src/builtin.rs:2937]:
```suggestion
                                    // Format arg labels are considered to be named labels for the purposes
```

File: compiler/rustc_lint/messages.ftl [hunk 6]
```diff
@@ -401,6 +397,19 @@ lint_incomplete_include =
 
 lint_inner_macro_attribute_unstable = inner macro attributes are unstable
 
+lint_invalid_asm_label_binary = avoid using labels containing only the digits 0 and 1 in inline assembly
```
Comment (estebank) [inline compiler/rustc_lint/messages.ftl:400]:
```suggestion
lint_invalid_asm_label_binary = avoid using labels containing only the digits `0` and `1` in inline assembly
```

File: tests/ui/asm/named-asm-labels.stderr [hunk 7]
```diff
@@ -245,111 +328,154 @@ LL |             ab: nop // ab: does foo
...
    |
    = help: only local labels of the form `<number>:` should be used in inline asm
+   = note: format arguments may expand to a non-numeric value
+   = note: see the asm section of Rust By Example <https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#labels> for more information
+
+error: avoid using named labels in inline assembly
+  --> $DIR/named-asm-labels.rs:156:14
+   |
+LL |         asm!(include_str!("named-asm-labels.s"));
+   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (estebank) [inline tests/ui/asm/named-asm-labels.stderr:424]: Interesting case. We should have a mechanism to display the contents of `include` and `include_str` and point *inside* of those files when complaining. At the very least, we should say "on line X col Y in file `foo.s`". Not to address in this PR.

File: compiler/rustc_lint/messages.ftl [hunk 8]
```diff
@@ -401,6 +397,19 @@ lint_incomplete_include =
...
+    .help = use a different number that has at least one digit 2 or greater
```
Comment (estebank) [inline compiler/rustc_lint/messages.ftl:401]: I'm unsure with the current wording. We mean that it can be *any* value other than 0 or 1, right?
```suggestion
    .label = use number 2 or greater
```

**Comment by joshtriplett** [comment]
On Tue, Jun 25, 2024 at 01:14:20PM -0700, Esteban Kuber wrote:
> +lint_invalid_asm_label_binary = avoid using labels containing only the digits 0 and 1 in inline assembly
> +    .help = use a different number that has at least one digit 2 or greater
> 
> I'm unsure with the current wording. We mean that it can be *any* value other than 0 or 1, right?
> ```suggestion
>     .label = use number 2 or greater
> ```

No, if I understand correctly, the problem is that any label whose
digits are only `0` and `1` can be confused with binary if written with
the 'b' suffix. This comes up most often with `0` and `1`, but if you
had a label `10` or `11` or `100` it would trigger the same bug. `10b`
to GCC is a reference to "label 10 backwards"; `10b` to
MASM/VC++/LLVM-in-intel-syntax-mode is `2` written in binary.

File: compiler/rustc_lint/messages.ftl (hunk 8 above)
Comment (asquared31415) [reply compiler/rustc_lint/messages.ftl:401]: `10:` is invalid, which is greater than 2.  A correct label specifically has at least one non-binary digit.  I'm not sure if there's a better way to put this.

File: tests/ui/asm/named-asm-labels.stderr (hunk 7 above)
Comment (asquared31415) [reply tests/ui/asm/named-asm-labels.stderr:424]: Agreed, I don't like the solution I came up with originally, but I opted not to change it for now, since that's a separate decision and complicates the implementation.

File: compiler/rustc_lint/src/builtin.rs (hunk 2 above)
Comment (asquared31415) [reply compiler/rustc_lint/src/builtin.rs:2827]: I forgot to expand this further, whoops.  I think I went to go get the bug link and got completely distracted along the way :sweat_smile: 

The Reference link below does also expand on it more and have the bug tracker link, but I think it's useful to duplicate a little bit of that here.

File: compiler/rustc_lint/src/builtin.rs (hunk 1 above)
Comment (asquared31415) [reply compiler/rustc_lint/src/builtin.rs:2832]: We do:
> Moreover, on x86 when using the default Intel syntax, due to [an LLVM bug](https://bugs.llvm.org/show_bug.cgi?id=36144), you shouldn't use labels exclusively made of 0 and 1 digits, e.g. 0, 11 or 101010, as they may end up being interpreted as binary values. Using options(att_syntax) will avoid any ambiguity, but that affects the syntax of the entire asm! block. (See [Options](https://doc.rust-lang.org/nightly/rust-by-example/unsafe/asm.html#options), below, for more on options.)

**Comment by asquared31415** [comment]
> This seems like a good starting point!
> 
> Eventually, I hope we can get LLVM to have a mode without this "feature", and then rust assembly should use that mode by default. But in the interim, we should warn people.

Somehow Clang does work around this.  Perhaps something that it does with using AT&T syntax by default and making the user write `.intel_syntax` is meaningful?  I don't know enough about how Clang uses inline assembly to be sure though.  Agreed that ideally this lint does get downgraded to allow by default by fixing the issue, but I think it's important in the meantime to give people a better error.

File: compiler/rustc_lint/messages.ftl (hunk 8 above)
Comment (estebank) [reply compiler/rustc_lint/messages.ftl:401]: I see, the issue is even more bizarre than I thought. In that case I'd reword this to

```suggestion
    .help = use a label that doesn't start with `0` or `1`
```

File: tests/ui/asm/named-asm-labels.rs [hunk 9]
```diff
@@ -141,7 +153,11 @@ fn main() {
         asm!("{1}: nop", "/* {0} */", const 10, const 20); //~ ERROR avoid using named labels
 
         // Test include_str in asm
-        asm!(include_str!("named-asm-labels.s")); //~ ERROR avoid using named labels
+        asm!(include_str!("named-asm-labels.s"));
+        //~^ ERROR avoid using named labels
+        //~^^ ERROR avoid using named labels
+        //~^^^ ERROR avoid using named labels
+        //~^^^^ ERROR avoid using named labels
```
Comment (estebank) [inline tests/ui/asm/named-asm-labels.rs:160]: For future reference, this can also be written as
```suggestion
        //~^ ERROR avoid using named labels
        //~| ERROR avoid using named labels
        //~| ERROR avoid using named labels
        //~| ERROR avoid using named labels
```

File: compiler/rustc_lint/messages.ftl (hunk 8 above)
Comment (asquared31415) [reply compiler/rustc_lint/messages.ftl]: It appears that `#[label]` only wants to be applied to a field of type `Span`. Is there a way to tell it to attach to the span that gets emitted with `cx.emit_span_lint(` ?

**Comment by estebank** [comment]
Please rebase on top of a recent HEAD and squash your commits. r=me after that.

**Comment by asquared31415** [comment]
I additionally implemented the requested change to make the suggestion a label on the span rather than a `help`.
